src/diagnostics/differences.py

```diff
@@ -0,0 +1,88 @@
+"""Functions to compute differences between treatment and control groups."""
+
+import torch
+import numpy as np
+import pandas as pd
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+def standarized_diffs(df_vars, treat_var, metric="smd", exclude=None, weights=None):
+    """Compute the standardized differences between the treatment and control groups.
+
+    To compare the quality of the weighting, we can check the standarized differences
+    between groups in our sample across the covariates used. This function calculates
+    to types of metrics: the standarized mean differences (SMD) and the absolute
+    standarized mean differences (ASMD). The function will return a dataframe with the
+    differences for each variable.
+
+    Args:
+        df_vars (pd.DataFrame): dataframe with the covariates
+        treat_var (str): name of the treatment variable
+        metric (str, optional): metric to use. Defaults to "smd".
+        exclude (list, optional): list of variables to exclude. Defaults to None.
+        weights (np.array, optional): weights to use in the calculation. Defaults to None.
+
+    Returns:
+        pd.DataFrame: dataframe with the standarized differences
+    """
+    # Split dataframe by treatment status
+
+    # Define metrics dict for conventions
+    metrics = {
+        "smd": "Standarized Mean Differences",
+        "asmd": "Absolute Standarized Mean Differences",
+    }
+
+    # Make the function work even if numpy arrays are passed directly rather than
+    # just a DataFrame
+    if isinstance(df_vars, pd.DataFrame):
+        X = df_vars.values
+        treat = df_vars[treat_var].values
+        index = df_vars.columns
+    else:
+        X = df_vars
+        treat = treat_var
+        index = [f"V{n}" for n in range(X.shape[1])]
+
+    # Transform to torch if not already
+    if not isinstance(X, torch.Tensor):
+        X = torch.from_numpy(X).float().to(device)
+        treat = torch.from_numpy(treat).float().to(device)
+
+    if weights is not None:
+        mean_treat = X[treat == 1, :].mean(axis=0)
+        mean_control = (X[treat == 0, :].T @ weights) / weights.sum()
+
+    else:
+        mean_treat = X[treat == 1, :].mean(axis=0)
+        mean_control = X[treat == 0, :].mean(axis=0)
+
+    # Calculate for all the variables
+    std_vars = X.std(axis=0)
```

Comment:
instead of pooling the variance across treatment and control, the advice in Imbens and Rubin is to compute $\sqrt{(\sigma^2_1 + \sigma^2_0)/2}$ where $\sigma^2_w$ is the variance of $X$ in treatment group $w$. This is particularly relevant in super asymmetrical T/C groups and lots of heteroskedasticity.

Reply:
did a bunch of other stuff but forgot to do this, haha. Lmk if you get around to it, otherwise i'll do it before merging.